spec/lib/connection_tester_spec.rb

```diff
@@ -0,0 +1,133 @@
+
+require "spec_helper"
+
+describe ConnectionTester do
+  describe "::check" do
+    it "takes a http url and returns a result object" do
+      res = ConnectionTester.check("https://pod.example.com")
+      expect(res).to be_a(ConnectionTester::Result)
+    end
+
+    it "still returns a result object, even for invalid urls" do
+      res = ConnectionTester.check("i:am/not)a+url")
+      expect(res).to be_a(ConnectionTester::Result)
+      expect(res.error).to be_a(ConnectionTester::Failure)
+    end
+  end
+
+  describe "#initialize" do
+    it "accepts the http protocol" do
+      expect {
+        ConnectionTester.new("https://pod.example.com")
+      }.not_to raise_error
+    end
+    it "rejects unexpected protocols" do
+      expect {
+        ConnectionTester.new("xmpp:user@example.com")
+      }.to raise_error(ConnectionTester::AddressFailure)
+    end
+  end
+
+  describe "#resolve" do
+    before do
+      @result = ConnectionTester::Result.new
+      @dns = instance_double("Resolv::DNS")
+      allow(@dns).to receive(:close).once
+    end
+
+    it "resolves the IP address" do
+      tester = ConnectionTester.new("https://pod.example.com", @result)
+      expect(tester).to receive(:with_dns_resolver).and_yield(@dns)
+      expect(@dns).to receive(:getaddress).and_return("192.168.1.2")
+
+      tester.resolve
+      expect(@result.ip).to eq("192.168.1.2")
+    end
+  end
+
+  describe "#request" do
+    before do
+      @url = "https://pod.example.com"
+      @stubs = Faraday::Adapter::Test::Stubs.new
+      @result = ConnectionTester::Result.new
+      @tester = ConnectionTester.new(@url, @result)
+
+      allow(@tester).to receive(:faraday_adapter).and_return([:test, @stubs])
```

Comment:
Why not just use webmock and save all that trouble (including having a hook method)?